functions/src/modules/gmail.functions.ts

```diff
@@ -0,0 +1,273 @@
+    const decodedMessage = JSON.parse(
+      Buffer.from(message.data, 'base64').toString('ascii')
+    );
+
+    // Gmail client and authenticate
+    const { client_key, secret, redirect } = functions.config().gmail;
+    const oauth2Client = new Google.google.auth.OAuth2(
+      client_key,
+      secret,
+      redirect
+    );
+    const { oauth, lastSyncHistoryId } = await fetchToken(
+      decodedMessage.emailAddress
+    );
+    oauth2Client.setCredentials({
+      access_token: oauth.token,
+      refresh_token: oauth.refreshToken,
+    });
+    const gm = await Google.google.gmail({
+      version: 'v1',
+      auth: oauth2Client,
+    });
+
+    const labelResults = await gm.users.labels.list({
+      userId: 'me',
+    });
+    const doneLabelObj = labelResults.data.labels.filter(label => {
+      return label.name === 'SQRDone';
+    })[0];
+
+    try {
+      // Typescript complains if not initialized and casted to any
+      const historyListRequest: any = {
+        userId: decodedMessage.emailAddress || 'me',
+        startHistoryId: lastSyncHistoryId,
+        historyTypes: 'labelAdded',
+        labelId: [doneLabelObj.id],
+      };
+
+      // Call the history list api to fetch changes since last synced
+      const results = await gm.users.history.list(historyListRequest);
+      if (!results.data.history || !results.data.history.length) {
+        // Handle empty history, and update history id
+        await storeHistoryIdInDatabase(
+          decodedMessage.emailAddress,
+          decodedMessage.historyId
+        );
+        throw new Error('History is empty. Nothing to process');
+      }
+
+      const onlyLabelAdded = results.data.history.filter(change => {
+        return change.labelsAdded;
+      });
+
+      // We only care about the final state of the Done label whether it was applied
+      // or not, so we get rid of the duplicates incase there are any
+      const uniques = {};
+      onlyLabelAdded.forEach(change => {
+        change.labelsAdded.forEach(obj => {
+          uniques[obj.message.id] = {
+            labels: obj.labelIds,
+            threadId: obj.message.threadId,
+          };
+        });
+      });
+
+      // Call the gmail.users.threads api with the unique changes
+      // and retrieve list of threads and messages
+      const threadHeaders = Object.keys(uniques).map(async threadKey => {
+        const threadResults = await gm.users.threads.get({
+          userId: 'me',
+          id: uniques[threadKey].threadId,
+        });
+        // Only care about the first message headers in the thread
+        return threadResults.data.messages[0].payload.headers;
+      });
+      const allResults = await Promise.all(threadHeaders);
+      // Filter the subject header out from the response
+      const subjectParts = allResults.map(thread => {
+        const subjectValue = thread.filter(
+          header => header.name === 'Subject'
+        )[0].value;
+        if (subjectValue) {
```

Comment:
Somewhere here you need to categorize the submission.
In the current task you are processing only SQR submissions.

> Subject pattern: `SQR`.

This means that only threads with `SQR` in the subject should be processed. You are not testing for this here.

Moreover, I propose to follow Separation of Concerns pattern.

The Gmail module should just categorize the submission by phase (CR, SQR, QC etc) and call corresponding function from corresponding module. SQR-specific actions should be implemented in the SQR module. Currently only SQR submissions processing should be implemented.